src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -2612,110 +2741,290 @@ AdapterHandlerEntry* AdapterHandlerLibrary::get_adapter(const methodHandle& meth
+      AdapterHandlerEntry* buffered_entry = ArchiveBuilder::current()->get_buffered_addr(entry);
+      assert(buffered_entry != nullptr,"sanity check");
+
+      uint hash = fp->compute_hash();
+      u4 delta = _builder->buffer_to_offset_u4((address)buffered_entry);
+      _writer->add(hash, delta);
+      if (lsh.is_enabled()) {
+        address fp_runtime_addr = (address)buffered_fp + ArchiveBuilder::current()->buffer_to_requested_delta();
+        address entry_runtime_addr = (address)buffered_entry + ArchiveBuilder::current()->buffer_to_requested_delta();
+        log_trace(cds)("Added fp=%p (%s), entry=%p to the archived adater table", buffered_fp, buffered_fp->as_basic_args_string(), buffered_entry);
+      }
+    } else {
+      if (lsh.is_enabled()) {
+        log_trace(cds)("Skipping adapter handler %p (fp=%s) as it is not archived", entry, fp->as_basic_args_string());
       }
     }
+    return true;
+  }
+};
+
+void AdapterHandlerLibrary::dump_aot_adapter_table() {
+  CompactHashtableStats stats;
+  CompactHashtableWriter writer(_adapter_handler_table->number_of_entries(), &stats);
+  CopyAdapterTableToArchive copy(&writer);
+  _adapter_handler_table->iterate(&copy);
+  writer.dump(&_aot_adapter_handler_table, "archived adapter table");
+}
+
+void AdapterHandlerLibrary::serialize_shared_table_header(SerializeClosure* soc) {
+  _aot_adapter_handler_table.serialize_header(soc);
+}
+
+AdapterBlob* AdapterHandlerLibrary::link_aot_adapter_handler(AdapterHandlerEntry* handler) {
+#ifdef ASSERT
+  if (TestAOTAdapterLinkFailure) {
+    return nullptr;
   }
 #endif
+  AdapterBlob* blob = lookup_aot_cache(handler);
+#ifndef PRODUCT
+  // debugging support
+  if ((blob != nullptr) && (PrintAdapterHandlers || PrintStubCode)) {
+    print_adapter_handler_info(handler, blob);
+  }
+#endif
+  return blob;
+}
 
-  // Add the entry only if the entry contains all required checks (see sharedRuntime_xxx.cpp)
-  // The checks are inserted only if -XX:+VerifyAdapterCalls is specified.
-  if (contains_all_checks || !VerifyAdapterCalls) {
-    assert_lock_strong(AdapterHandlerLibrary_lock);
-    _adapter_handler_table->put(fingerprint, entry);
+// This method is used during production run to link archived adapters (stored in AOT Cache)
+// to their code in AOT Code Cache
+void AdapterHandlerEntry::link() {
+  AdapterBlob* adapter_blob = nullptr;
+  ResourceMark rm;
+  assert(_fingerprint != nullptr, "_fingerprint must not be null");
+  bool generate_code = false;
+  // Generate code only if AOTCodeCache is not available, or
+  // caching adapters is disabled, or we fail to link
+  // the AdapterHandlerEntry to its code in the AOTCodeCache
+  if (AOTCodeCache::is_using_adapters()) {
+    adapter_blob = AdapterHandlerLibrary::link_aot_adapter_handler(this);
+    if (adapter_blob == nullptr) {
+      log_warning(cds)("Failed to link AdapterHandlerEntry (fp=%s) to its code in the AOT code cache", _fingerprint->as_basic_args_string());
```

Comment:
Doesn't it add noise in the output for not yet seen adapter shapes? It's a warning.

Reply:
`AdapterHandlerEntry::link` method gets called only for the archived adapters. Not yet seen adapters do not come across this code path. Reason for making it a warning is the assumption that, if the AOT Code Cache is usable, then we should always be able to link the `AdapterHandlerEntry` to its code in the AOT code cache. If we fail to do so, then something is not right in this machinery and likely pointing to a bug.